src/Nethermind/Nethermind.Optimism/CL/L1Bridge/EthereumL1Bridge.cs

```diff
@@ -0,0 +1,293 @@
+    {
+        Hash256 currentHash = headHash;
+        L1Block[] chainSegment = new L1Block[headNumber - segmentStartNumber + 1];
+        for (int i = chainSegment.Length - 1; i >= 0; i--)
+        {
+            _logger.Info($"Rolling back L1 head. {i} to go. Block {currentHash}");
+            chainSegment[i] = await GetBlockByHash(currentHash);
+            currentHash = chainSegment[i].ParentHash;
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return;
+            }
+        }
+
+        for (int i = 0; i < chainSegment.Length; i++)
+        {
+            await ProcessBlock(chainSegment[i], cancellationToken);
+        }
+    }
+
+    // Gets all blocks from range ({from}, {to}). It's safe only if {to} is finalized
+    private async Task BuildUp(ulong from, ulong to, CancellationToken cancellationToken)
+    {
+        for (ulong i = from + 1; i < to; i++)
+        {
+            await ProcessBlock(await GetBlock(i), cancellationToken);
+        }
+    }
+
+    public async Task<L1Block> GetBlock(ulong blockNumber)
+    {
+        L1Block? result = await _ethL1Api.GetBlockByNumber(blockNumber, true);
+        while (result is null)
+        {
+            if (_logger.IsWarn) _logger.Warn($"Unable to get L1 block by block number({blockNumber})");
+            result = await _ethL1Api.GetBlockByNumber(blockNumber, true);
+        }
+
+        return result.Value;
+    }
+
+    public async Task<L1Block> GetBlockByHash(Hash256 blockHash)
+    {
+        L1Block? result = await _ethL1Api.GetBlockByHash(blockHash, true);
+        while (result is null)
+        {
+            if (_logger.IsWarn) _logger.Warn($"Unable to get L1 block by hash({blockHash})");
+            result = await _ethL1Api.GetBlockByHash(blockHash, true);
+        }
+
+        return result.Value;
+    }
+
+    public async Task<ReceiptForRpc[]> GetReceiptsByBlockHash(Hash256 blockHash)
+    {
+        ReceiptForRpc[]? result = await _ethL1Api.GetReceiptsByHash(blockHash);
+        while (result is null)
+        {
+            if (_logger.IsWarn) _logger.Warn($"Unable to get L1 receipts by hash({blockHash})");
+            result = await _ethL1Api.GetReceiptsByHash(blockHash);
+        }
+
+        return result;
+    }
+
+    private async Task<L1Block> GetHead()
+    {
+        L1Block? result = await _ethL1Api.GetHead(true);
+        while (result is null)
+        {
+            _logger.Warn($"Unable to get L1 head");
+            result = await _ethL1Api.GetHead(true);
+        }
+
+        return result.Value;
+    }
+
+    private async Task<L1Block> GetFinalized()
+    {
+        L1Block? result = await _ethL1Api.GetFinalized(true);
+        while (result is null)
+        {
+            _logger.Warn($"Unable to get finalized L1 block");
+            result = await _ethL1Api.GetFinalized(true);
+        }
+
+        return result.Value;
+    }
+
+    public void Reset(L1BlockInfo highestFinalizedOrigin)
```

Comment:
Reset is used only once by `OptimismCL` during initialization (`.Start()`).
Would it be possible to move this code into the constructor, removing this method and avoiding possible race conditions (ex. calling `Reset` while `Run()` is still active)?

Reply:
It will be used later to reset pipeline in case of reorgs

Reply:
Not rn at least